### PR TITLE
Allow Bottom element kind for empty arrays

### PR DESCRIPTION
**Pull request by lthls**
This should fix the errors with empty arrays used at different types.

### CODE REVIEW

File: middle_end/flambda2/types/meet_and_join.ml [hunk 1]
```diff
@@ -386,11 +390,13 @@ and meet_head_of_kind_value env (head1 : TG.head_of_kind_value)
           alloc_mode = alloc_mode2
         } ) ->
     let<* alloc_mode = meet_alloc_mode alloc_mode1 alloc_mode2 in
-    let<* element_kind = meet_array_element_kinds element_kind1 element_kind2 in
+    let element_kind = meet_array_element_kinds element_kind1 element_kind2 in
     let<* contents, env_extension =
       meet_array_contents env array_contents1 array_contents2
     in
     let<* length, env_extension' = meet env length1 length2 in
+    (* Note: If the element kind is Bottom, we could meet the length type with
```
Comment (mshinwell) [inline middle_end/flambda2/types/meet_and_join.ml:398]: This should probably be a `CR-someday`.  Missing parenthesis at the end of the comment too.

File: middle_end/flambda2/types/reify.ml [hunk 2]
```diff
@@ -384,7 +384,10 @@ let reify ~allowed_if_free_vars_defined_in ~var_is_defined_at_toplevel
       | _ :: _ -> (
         match element_kind with
         | Unknown -> try_canonical_simple ()
-        | Known element_kind -> (
+        | Bottom ->
+          (* Note: we could use [Lift Empty_array] here *)
```
Comment (mshinwell) [inline middle_end/flambda2/types/reify.ml:388]: Same again, make this a `CR-someday` as a potential future improvement.

File: middle_end/flambda2/types/provers.ml [hunk 3]
```diff
@@ -495,7 +495,10 @@ let meet_is_flat_float_array env t : bool meet_shortcut =
   | Value Unknown -> Need_meet
   | Value Bottom -> Invalid
   | Value (Ok (Array { element_kind = Unknown; _ })) -> Need_meet
-  | Value (Ok (Array { element_kind = Known element_kind; _ })) -> (
+  | Value (Ok (Array { element_kind = Bottom; _ })) ->
+    (* Empty array case. We cannot return Bottom, but any result is correct. *)
```
Comment (mshinwell) [inline middle_end/flambda2/types/provers.ml:499]: True, but returning `false` here seems more consistent with the fact that the empty array always has tag zero, right?

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/types/provers.ml (hunk 3 above)
Comment (lthls) [reply middle_end/flambda2/types/provers.ml:499]: Actually I'm a bit worried about the code in `Simplify_common.specialise_array_kind`. I may have to return `Need_meet` just to make sure that we're not turning valid array operations on an empty array into `Invalid`.

File: middle_end/flambda2/types/meet_and_join.ml (hunk 1 above)
Comment (lthls) [reply middle_end/flambda2/types/meet_and_join.ml]: Done

File: middle_end/flambda2/types/reify.ml (hunk 2 above)
Comment (lthls) [reply middle_end/flambda2/types/reify.ml]: Done

File: middle_end/flambda2/types/provers.ml (hunk 3 above)
Comment (lthls) [reply middle_end/flambda2/types/provers.ml]: I've expanded the comment here a bit, and added comments at the use points too.

**Review by mshinwell — APPROVED** [review]

**Pull request merged**